File: achievement-module/achievement-fee/src/main/java/com/institute/achievement/fee/entity/FeeRecord.java
package com.institute.achievement.fee.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Fee record entity — maps to the {@code fee_record} table.
 * <p>
 * Records fees associated with patents and software copyrights via the
 * Arc polymorphic pattern ({@code ownerType} / {@code ownerId}).
 * Supports the full fee lifecycle: pending → paid | paused.
 *
 * @see com.institute.achievement.fee.enums.FeeTypeEnum
 * @see com.institute.achievement.fee.enums.FeeStatusEnum
 */
@Data
@TableName("fee_record")
public class FeeRecord {

    @TableId(type = IdType.AUTO)
    private Long id;

    // ── Fee Type ──────────────────────────────────────────────────────

    /** Fee type code — maps to {@link com.institute.achievement.fee.enums.FeeTypeEnum} */
    @TableField("fee_type")
    private String feeType;

    // ── Amounts ───────────────────────────────────────────────────────

    /** Total fee amount (yuan) */
    private BigDecimal amount;

    /** Actual paid amount (yuan) — may differ from amount if partial payment */
    @TableField("paid_amount")
    private BigDecimal paidAmount;

    // ── Dates ─────────────────────────────────────────────────────────

    /** Fee due date */
    @TableField("due_date")
    private LocalDate dueDate;

    /** Actual payment date */
    @TableField("paid_date")
    private LocalDate paidDate;

    // ── Payment Evidence ──────────────────────────────────────────────

    /** Payment voucher / receipt number */
    @TableField("voucher_no")
    private String voucherNo;

    // ── Status ────────────────────────────────────────────────────────

    /** Fee status code — maps to {@link com.institute.achievement.fee.enums.FeeStatusEnum} */
    private String status;

    // ── Financial ─────────────────────────────────────────────────────

    /** Funding source code — configured via data dictionary (FUND_SOURCE) */
    @TableField("funding_source")
    private String fundingSource;

    // ── Polymorphic Arc (D-10) ────────────────────────────────────────

    /** Owner type — "patent" or "copyright" */
    @TableField("owner_type")
    private String ownerType;

    /** Owner entity ID */
    @TableField("owner_id")
    private Long ownerId;

    // ── Source ────────────────────────────────────────────────────────

    /** Record source — "auto_generated" or "manual" */
    private String source;

    // ── Batch Slip (used in 02-04) ────────────────────────────────────

    /** Batch slip number, filled during batch slip generation */
    @TableField("slip_no")
    private String slipNo;

    /** Timestamp when the batch slip was generated */
    @TableField("slip_generated_time")
    private LocalDateTime slipGeneratedTime;

    /** User ID who generated the batch slip */
    @TableField("slip_generated_by")
    private Long slipGeneratedBy;

    // ── Department & Ownership ────────────────────────────────────────

    /** Department ID (for SQL-layer data isolation) */
    @TableField("dept_id")
    private Long deptId;

    /** Creator user ID */
    @TableField("created_by")
    private Long createdBy;

    @TableField(fill = FieldFill.INSERT)
    private LocalDateTime createdTime;

    /** Last updater user ID */
    @TableField("updated_by")
    private Long updatedBy;

    @TableField(fill = FieldFill.UPDATE)
    private LocalDateTime updatedTime;
}

File: achievement-web/src/api/fee/feeRecord.ts
import http from '@/api/index'

/**
 * Fee record API response interface matching backend FeeRecordVO.
 */
export interface FeeRecordVO {
  id: number
  feeType: string
  amount: number
  paidAmount?: number
  dueDate: string
  paidDate?: string
  voucherNo?: string
  status: string
  fundingSource?: string
  ownerType: string
  ownerId: number
  source: string
  slipNo?: string
  slipGeneratedTime?: string
  slipGeneratedBy?: number
  deptId?: number
  createdBy?: number
  createdTime?: string
  updatedBy?: number
  updatedTime?: string

  // Computed fields
  ownerName?: string
  alertLevel?: string
  statusLabel?: string
  feeTypeLabel?: string
}

export interface PageResult<T> {
  records: T[]
  total: number
  size: number
  current: number
  pages: number
}

/**
 * Fee record paginated listing params.
 */
export interface FeeRecordPageParams {
  page: number
  size: number
  status?: string
  feeType?: string
  fundingSource?: string
  keyword?: string
  dueDateFrom?: string
  dueDateTo?: string
  ownerType?: string
}

/**
 * Fee record create/update DTO.
 */
export interface FeeRecordDTO {
  feeType: string
  amount: number
  paidAmount?: number
  dueDate: string
  voucherNo?: string
  status?: string
  fundingSource?: string
  ownerType: string
  ownerId: number
}

/**
 * Get paginated fee records with multi-dimensional filtering.
 */
export function getPage(params: FeeRecordPageParams) {
  return http.get('/api/fees/page', { params })
}

/**
 * Get a fee record by ID.
 */
export function getById(id: number) {
  return http.get(`/api/fees/${id}`)
}

/**
 * Create a new fee record.
 */
export function create(data: FeeRecordDTO) {
  return http.post('/api/fees', data)
}

/**
 * Update an existing fee record.
 */
export function update(id: number, data: FeeRecordDTO) {
  return http.put(`/api/fees/${id}`, data)
}

/**
 * Delete a fee record (only allowed for paused records).
 */
export function remove(id: number) {
  return http.delete(`/api/fees/${id}`)
}

// ── Batch Payment Operations (02-04 Task 3) ──────────────────────────

/**
 * Batch generate slip numbers for selected pending fee records.
 */
export function batchGenerateSlips(ids: number[]) {
  return http.post('/api/fees/batch-generate-slips', { ids })
}

/**
 * Batch mark fee records as paid.
 */
export function batchPay(data: { ids: number[], paidDate: string, voucherNo: string, slipNo: string }) {
  return http.put('/api/fees/batch-pay', data)
}
